src/util/boolset.rs

```diff
@@ -0,0 +1,78 @@
+use std::cmp::Eq;
+use std::collections::HashMap;
+use std::default::Default;
+use std::fmt::Debug;
+use std::hash::Hash;
+
+// Like a hash table, the bool set is a data structure where keys
+// can be explicitly set to true or false.
+
+#[derive(Debug, Default, Clone, PartialEq)]
+pub struct BoolSet<T>
+where
+    T: Hash + Eq + Default + Clone,
+{
+    map: HashMap<T, bool>,
+    default_val: bool,
+}
+
+impl<T> BoolSet<T>
+where
+    T: Hash + Eq + Default + Clone,
+{
+    pub fn new_with_default(default_val: bool) -> Self {
+        BoolSet {
+            map: HashMap::new(),
+            default_val,
+        }
+    }
+
+    // Maps the key to the value in bool set, overwriting any existing mapping for the key.
+    pub fn insert(&mut self, key: T, val: bool) {
+        self.map.insert(key, val);
+    }
+
+    // Returns the value for the key in bool set.
+    // If no value is found for the key, then default value is returned.
+    pub fn get(&self, key: &T) -> bool {
```

Comment:
This should be a similar signature than HashMap::get, so:

```
pub fn get<Q>(&self, key: &Q) -> bool
where
    K: Borrow<Q>,
    Q: Hash + Eq,
{
```

That way you can call it with static strings when called on strings like:

```
set.get("string")
```

Rather than:

```
set.get(&"string".to_owned())
```